src/metabase/models/task_history.clj

```diff
@@ -71,45 +87,53 @@
 (def ^:private TaskHistoryInfo
   "Schema for `info` passed to the `with-task-history` macro."
   [:map {:closed true}
-   [:task                          ms/NonBlankString] ; task name, i.e. `send-pulses`. Conventionally lisp-cased
-   [:db_id        {:optional true} [:maybe :int]]     ; DB involved, for sync operations or other tasks where this is applicable.
-   [:task_details {:optional true} [:maybe :map]]])   ; additional map of details to include in the recorded row
-
-(defn- save-task-history! [start-time-ms info]
-  (let [end-time-ms (System/currentTimeMillis)
-        duration-ms (- end-time-ms start-time-ms)]
-    (try
-      (first (t2/insert-returning-instances! TaskHistory
-                                             (assoc info
-                                                    :started_at (t/instant start-time-ms)
-                                                    :ended_at   (t/instant end-time-ms)
-                                                    :duration   duration-ms)))
-      (catch Throwable e
-        (log/warn e "Error saving task history")))))
+   [:task                             ms/NonBlankString] ; task name, i.e. `send-pulses`. Conventionally lisp-cased
+   [:db_id           {:optional true} [:maybe :int]]     ; DB involved, for sync operations or other tasks where this is applicable.
+   ;; a function that takes the result of the task and returns a map of additional info to update task history when the task succeeds
+   [:on-success-info {:optional true} [:maybe [:=> [:cat :any] :map]]]
+   [:task_details    {:optional true} [:maybe :map]]])   ; additional map of details to include in the recorded row
+
+(defn- update-task-history!
+  [th-id startime-ms info]
+  (let [end-time-ms  (System/currentTimeMillis)
+        updated-info (merge {:ended_at (t/instant end-time-ms)
+                             :duration (- end-time-ms startime-ms)}
+                            info)]
```

Comment:
Chrisbot[^1]: always use `System/nanoTime` for measuring durations.

I guess this is a bit of a grey area, as you *don't* want to use that for `:ended_at`. I would just use `(t/instant)` directly.

[^1]: maybe I need to write a linter for this